CMSgov/design-system
# [WNMGDS-2992] Allow pass through props into TextField from ds-autocomplete

## Summary

- [Ticket](https://jira.cms.gov/browse/WNMGDS-2992)
- Initially this ticket was set to test the feasibility of adding a `ds-text-field` web component as a child element of the `ds-autocomplete` component; while possible setting a `ds-text-field` component as a child presents some hairy difficulties:
1. Conflict between the text field's `onChange` event and the `ds-autocomplete`'s event
2. Issues with onFocus and onBlur being called simultaneously - tabbing to the autocomplete would call both, and then a second tab press would be need to focus the underlying input
3. It's unclear if the correct DOM nodes were being grabbed by the `ds-autocomplete` element from the child `ds-text-field`. As far as I could tell, the `ds-autocomplete` element was consuming the HTML from within the child <template> rather than the rendered HTML. 

Given these difficulties, we opted to pursue expanding the input attributes to the underlying `TextField` component in the `ds-autocomplete`. 

Many of these fields are composed from other fields passed into the `ds-autocomplete`, and several fields should be left off to preserve a usable UI. 

What I did is examine the props passed to the `TextField` component in the React `Autocomplete` component and determine which of those should be exposed to users of the web component version of the Autocomplete. It's not many.

New pass through attributes:
1. errorMessageClassName
2. errorPlacement
3. errorMessage
4. disabled - this one I'm happy to exclude, but I know our guidance around disabled UI elements is forthcoming.

Existing attributes:
1. value
2. name - this should always be 'autocomplete'
3. label

Omitted attributes:
1. autoComplete - composed from props passed to Autocomplete
2. autoFocus - composed from props passed to Autocomplete
3. aria-activedescendant - I don't think this should be set by a user, but rather programmatically
4. aria-controls - composed from props passed to Autocomplete
5. aria-labelledby - comes from the `labelId` attribute on the TextField
6. id - composed from props passed to Autocomplete
7. size - changing the size from 'full' to 'medium' or 'small' is not good from a usability stand point. Making the input smaller crowds the items that appear in the drop down.

## How to test

1. Run local tests
2. Change some of the new pass through attributes in Storybook and see that the changes appear in the rendered result.

## Checklist
